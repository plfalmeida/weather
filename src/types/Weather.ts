
export interface WeatherInfo {
  coord:      Coord;
  weather:    Weather[];
  base:       string;
  main:       Main;
  visibility: number;
  wind:       Wind;
  clouds:     Clouds;
  dt:         number;
  sys:        Sys;
  timezone:   number;
  id:         number;
  name:       string;
  cod:        number;
}

export interface Clouds {
  all: number;
}

export interface Coord {
  lon: number;
  lat: number;
}

export interface Main {
  temp:       number;
  feels_like: number;
  temp_min:   number;
  temp_max:   number;
  pressure:   number;
  humidity:   number;
}

export interface Sys {
  type:    number;
  id:      number;
  country: string;
  sunrise: number;
  sunset:  number;
}

export interface Weather {
  id:          number;
  main:        string;
  description: string;
  icon:        string;
}

export interface Wind {
  speed: number;
  deg:   number;
}

export interface WeatherState {
  cities: string[];
  currentWeather: WeatherInfo;
  // TODO: Add correct type
  hourlyDetail: {
    hourly:   any[];
    timezone: string;
  };
  locale: string;
}

export interface TemperatureInfo {
  current: number;
  max:     number;
  min:     number;
}

export interface DayInfo {
  name:  string;
  value: number;
}
