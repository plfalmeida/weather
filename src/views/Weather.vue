<template lang="pug">

  .weather(:data-theme="weather.main")
    nav.top-navigation
      router-link(:to="{ name: 'home' }").back-button
        arrow-icon(direction="left")
    .container
      header
        h1 {{ city }}
        h2 {{ weather.main }}

      main(@click="changeIconSet")
        temperature(:temperature="temperature")
        weather-icon(:iconId="weather.icon" iconSize="@4x" :custom="showCustomIcons")
        temperature-extended(:periods="hourly" :customIcons="showCustomIcons")
        day-info(:dailyInfo="dailyInfo")

</template>

<script>
import { Vue, Component } from 'vue-property-decorator'

import Temperature from '@/components/Temperature/Temperature.vue'
import TemperatureExtended from '@/components/Temperature/TemperatureExtended.vue'
import DayInfo from '@/components/Day/DayInfo.vue'
import WeatherIcon from '@/components/ui/WeatherIcon.vue'
import ArrowIcon from '../components/ui/ArrowIcon.vue'

import { mapGetters } from 'vuex'

@Component({
  name: 'weather',
  components: {
    Temperature,
    TemperatureExtended,
    DayInfo,
    WeatherIcon,
    ArrowIcon,
  },
  computed: {
    ...mapGetters(['temperature', 'dailyInfo', 'weather', 'city', 'hourly', 'showCustomIcons']),
  },
})
export default class Weather extends Vue {
  mounted () {
    const { city } = this.$route.params
    this.$store.dispatch('fetchWeather', { cityName: city })
  }

  changeIconSet () {
    this.$store.commit('toggleIconSet')
  }
}
</script>

<style lang="stylus" scoped>

header
  margin-bottom 2rem

h1
  font-size 4.5rem
  text-transform uppercase

h2
  margin-top 0.5rem
  font-size 2.5rem
  font-weight 200
  text-transform lowercase

.top-navigation
  width 100%
  position fixed
  top 0
  left 0
  text-align left

.back-button
  padding 1rem
  display inline-block
  >>> svg
    width 2rem

.weather
  padding 5rem 0
  min-height 100vh
  overflow-y auto
  color var(--w-foreground)
  background var(--w-background)
  text-align center
  display flex
  justify-content center
  transition all 0.8s

  &[data-theme="Clear"]
    --w-background linear-gradient(to bottom, #00c6ff, #0072ff)

  &[data-theme="Clouds"]
    --w-background linear-gradient(to bottom, #00c6ff, #b0bac5)

  &[data-theme="Drizzle"]
    --w-background linear-gradient(to bottom, #a7bece, #6aa4de)

  &[data-theme="Rain"]
    --w-background linear-gradient(to bottom, #bdc3c7, #2c3e50)

  &[data-theme="Thunderstorm"]
    --w-background linear-gradient(to bottom, #8e5da0, #1a0540)

  &[data-theme="Snow"]
    --w-background linear-gradient(to bottom, #f8fcff, #d0d0d0)
    --w-foreground #1d2d3e

  @media screen and (min-height: 650px)
    align-items center
</style>
