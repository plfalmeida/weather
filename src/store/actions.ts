import weatherService from '@/services/weatherService'
import { ActionTree } from 'vuex'
import { WeatherState } from '@/types/Weather'

const actions: ActionTree<WeatherState, WeatherState> = {

  fetchWeather: async (context, { cityName }: { cityName: string }) => {
    try {
      const { data } = await weatherService.findByCity(cityName)
      localStorage.setItem('weather', JSON.stringify(data))
      context.commit('setWeather', data)
      context.dispatch('fetchHourly', {
        lat: data.coord.lat,
        lon: data.coord.lon,
      })
    } catch (err) {
      console.log('err', err)
    }
  },

  fetchHourly: async (context, { lat, lon }) => {
    try {
      const { data } = await weatherService.findHourly(lat, lon)
      localStorage.setItem('hourly', JSON.stringify(data))
      context.commit('setHourly', data)
    } catch (err) {
      console.log('err', err)
    }
  },

}

export default actions
