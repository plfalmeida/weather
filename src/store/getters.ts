import { WeatherState } from '@/types/Weather'
import { GetterTree } from 'vuex'
import { unixToDate, zoneDateFromUnix, getHour } from '@/helpers/date'

const getters: GetterTree<WeatherState, WeatherState> = {
  temperature: (state) => {
    let { temp: current, temp_max: max, temp_min: min } = state.currentWeather.main

    current = Math.trunc(current)
    max = Math.trunc(max)
    min = Math.trunc(min)

    return { current, max, min }
  },

  dailyInfo: (state) => {
    const { locale } = state
    const { timezone } = state.hourlyDetail

    const speed = state.currentWeather.wind.speed
    const humidity = state.currentWeather.main.humidity

    const sunrise = getHour(
      unixToDate(state.currentWeather.sys.sunrise),
      timezone,
      locale,
    )

    const sunset = getHour(
      unixToDate(state.currentWeather.sys.sunset),
      timezone,
      locale,
    )

    return [
      { name: 'windspeed', value: `${speed} m/s` },
      { name: 'sunrise', value: sunrise },
      { name: 'sunset', value: sunset },
      { name: 'humidity', value: `${humidity}%` },
    ]
  },

  weather: (state) => {
    const [CurrentWeather = { main: '' }] = state.currentWeather.weather
    return CurrentWeather
  },

  city: (state) => state.currentWeather.name,

  hourly: (state) => {
    type period = {
      index: number;
      name: string;
    }

    const periods: { [key: string]: period } = {
      '3:00 AM': {
        index: 1,
        name: 'Dawn',
      },
      '9:00 AM': {
        index: 2,
        name: 'Morning',
      },
      '3:00 PM': {
        index: 3,
        name: 'Afternoon',
      },
      '9:00 PM': {
        index: 4,
        name: 'Night',
      },
    }

    const { locale } = state
    const { timezone, hourly } = state.hourlyDetail
    const relevantTimes = Object.keys(periods)

    return hourly.reduce((acc: any[], hour: any) => {
      const date = zoneDateFromUnix(hour.dt, timezone)
      hour.time = getHour(date, timezone, locale)

      if (relevantTimes.includes(hour.time)) {
        hour.temp = Math.trunc(hour.temp)
        const [weather] = hour.weather
        hour.weather = weather
        hour.name = periods[hour.time].name
        hour.index = periods[hour.time].index
        acc.push(hour)
      }

      return acc
    }, []).sort((a, b) => (a.index > b.index) ? 1 : -1)
  },
}

export default getters
