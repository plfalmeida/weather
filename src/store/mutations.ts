import { WeatherState } from '@/types/Weather'
import { MutationTree } from 'vuex'

const mutations: MutationTree<WeatherState> = {
  setWeather: (state, payload) => {
    state.currentWeather = payload
  },
  setHourly: (state, payload) => {
    payload.hourly.length = 23
    state.hourlyDetail = payload
  },
}

export default mutations
