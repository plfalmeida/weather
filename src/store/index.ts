import Vue from 'vue'
import Vuex from 'vuex'
import { WeatherState } from '@/types/Weather'

import actions from './actions'
import getters from './getters'
import mutations from './mutations'

Vue.use(Vuex)

const state = {
  currentWeather: {
    main: {},
    wind: {},
    sys: {},
    weather: [],
  },
  hourlyDetail: {
    hourly: [],
  },
  locale: navigator.language,
  cities: [
    'Dallol',
    'Fairbanks',
    'London',
    'Recife',
    'Vancouver',
    'Yakutsk',
  ],
} as unknown as WeatherState

export default new Vuex.Store({
  state,
  actions,
  getters,
  mutations,
})
