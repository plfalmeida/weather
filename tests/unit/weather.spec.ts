import { shallowMount, createLocalVue } from '@vue/test-utils'
import Weather from '@/views/Weather.vue';

import VueRouter from 'vue-router'
import Vuex from 'vuex'

import store from '../../src/store';


const localVue = createLocalVue()

localVue.use(VueRouter)
localVue.use(Vuex)

const router = new VueRouter()

describe('Weather.vue', ()=> {
  it('Match snapshot', () => {
    const wrapper = shallowMount(Weather, { localVue, router, store })
    expect(wrapper).toMatchSnapshot()
  })
})
