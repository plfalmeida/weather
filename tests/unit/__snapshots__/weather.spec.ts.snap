// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather.vue Match snapshot 1`] = `
<div data-theme="" class="weather">
  <div class="container">
    <header>
      <h1></h1>
      <h2></h2>
    </header>
    <main>
      <temperature-stub temperature="[object Object]"></temperature-stub>
      <weather-icon-stub iconid="02d" iconsize="@4x"></weather-icon-stub>
      <temperature-extended-stub periods=""></temperature-extended-stub>
      <day-info-stub dailyinfo="[object Object],[object Object],[object Object],[object Object]"></day-info-stub>
    </main>
  </div>
</div>
`;
